src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupStartupFilter.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+using System.Text;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.AutoSetup.Options;
+using OrchardCore.Environment.Shell.Scope;
+
+namespace OrchardCore.AutoSetup
+{
+    /// <summary>
+    /// The filter which registers <see cref="AutoSetupMiddleware"/> to setup the site.
+    /// </summary>
+    public class AutoSetupStartupFilter : IStartupFilter
```

Comment:
Is it necessary since you could register the middleware from Configure in the same Startup class that registers this filter?

Reply:
Thanks for review
I thought about this, I think I could do this by using middleware in the Startup. I guess the main difference is that ASP.NET Core uses IStartupFilters in order to build the application pipeline, by using IStartupFilter I can guarantee that AutosetupMiddleware will be the first in the pipeline. Please correct me if I'm wrong.

Reply:
Yes, your startup filter will run first (app level ones still run first) as we call the startup filters of a given tenant (using its own app builder) before building its pipeline (so calling the modules startup `Configure()`).

But for testing you can try to use a startup class, the one related to your auto setup feature, and override the `ConfigureOrder` property with a negative value, e.g. -100

But just to test it because somewhere i like the idea of using a startup filter, the only thing is that the startup `Configure()` provides the shell scope service provider, as commented in another comment, but for me not an issue by retrieving it with any other way

Reply:
Thanks for the reviews, I replaced AutoSetupStartupFilter  as a middleware in Statup.cs